[Fix][Connector-V2] Fixed clickhouse connectors cannot stop under multiple parallelism

YOMO-Lee opened the pull request:
Fixed the bug where clickhouse connectors cannot stop properly under multiple parallelism

YOMO-Lee committed:
[Fix][Doc] Fix LocalFile doc (#7887)

Supplement and optimize the description of the LocalFile connector on filtering files
[(#7887)](https://github.com/apache/seatunnel/issues/7887)

YOMO-Lee committed:
Merge branch 'apache:dev' into dev

YOMO-Lee committed:
Merge branch 'apache:dev' into dev

YOMO-Lee committed:
[Fix][Connector-V2][ClickHouse] Fix ClickHouse Bug (#7897)

1、When the ClickHouse connector is set to multi parallelism, the task extraction is completed but cannot be stopped normally
[(#7897)](https://github.com/apache/seatunnel/issues/7897)

2、Added E2E test cases for this issue [(#7897)](https://github.com/apache/seatunnel/issues/7897)

3、Local developers want to observe **Job Progress Information** in a timely manner,  Need to modify the following configuration.The configuration in config is invalid
```
seatunnel engine/seatunnel-engineer-common/src/main/resources/seatunnely.yaml
```

YOMO-Lee committed:
[Fix][Connector-V2][ClickHouse] Fix ClickHouse Bug (#7897)

1、When the ClickHouse connector is set to multi parallelism, the task extraction is completed but cannot be stopped normally
[(#7897)](https://github.com/apache/seatunnel/issues/7897)

2、Added E2E test cases for this issue [(#7897)](https://github.com/apache/seatunnel/issues/7897)

3、Local developers want to observe **Job Progress Information** in a timely manner,  Need to modify the following configuration.The configuration in config is invalid
```
seatunnel engine/seatunnel-engineer-common/src/main/resources/seatunnely.yaml
```

YOMO-Lee committed:
[Fix][Connector-V2][ClickHouse] Fix ClickHouse Bug (#7897)

1、When the ClickHouse connector is set to multi parallelism, the task extraction is completed but cannot be stopped normally
[(#7897)](https://github.com/apache/seatunnel/issues/7897)

2、Added E2E test cases for this issue [(#7897)](https://github.com/apache/seatunnel/issues/7897)

3、Local developers want to observe **Job Progress Information** in a timely manner, Need to modify the following configuration.The configuration in config is invalid
```
seatunnel engine/seatunnel-engineer-common/src/main/resources/seatunnely.yaml
```

YOMO-Lee committed:
[Fix][Doc] Fix LocalFile doc (#7887)

Continue to optimize the document about filtering files and add some examples
[(#7887)](https://github.com/apache/seatunnel/issues/7887)

YOMO-Lee committed:
Merge branch 'apache:dev' into dev

YOMO-Lee committed:
[Fix][Doc] Fix LocalFile doc (#7887)

Change to English log
[(#7887)](https://github.com/apache/seatunnel/issues/7887)

YOMO-Lee committed:
Merge remote-tracking branch 'origin/dev' into dev

YOMO-Lee committed:
Merge branch 'apache:dev' into dev

YOMO-Lee committed:
Revert "[Fix][Doc] Fix LocalFile doc (#7887)"

This reverts commit 42e591925056841d1299effa0112d8134d5d3070.

YOMO-Lee committed:
Revert "[Fix][Doc] Fix LocalFile doc (#7887)"

This reverts commit baa7cb89e40c55159414cc4d4c9c6ad5b2875d82.

corgy-w commented on seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:
```diff
@@ -20,6 +20,7 @@ seatunnel:
         backup-count: 1
         queue-type: blockingqueue
         print-execution-info-interval: 60
+        print-job-metrics-info-interval: 60
```
Default value of `print-job-metrics-info-interval` is 60. This is valid without change

YOMO-Lee replied:
I know that the default value is also used if it is not changed here, but here is a parameter to let you know where you can change it.
Otherwise, if you run an example in the project and want to change the parameters, the results will not work

YOMO-Lee replied:
> Default value of `print-job-metrics-info-interval` is 60. This is valid without change

I think this parameter can be added here, at least adding the default value will not cause any other bad effects
But it worked for me, at first I wanted to change the configuration, I didn't know where to change, I searched a lot of places in the project, and it didn't take effect after the change, and then it was someone else who pointed out that I knew to change here

corgy-w replied:
get

Hisoka-X replied:
You can configure it in your local environment. But do not commit it. Please revert it.

Hisoka-X commented on seatunnel-e2e/seatunnel-connector-v2-e2e/connector-clickhouse-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/ClickhouseIT.java:
```diff
@@ -101,6 +101,13 @@ public void testClickhouse(TestContainer container) throws Exception {
 
+    @TestTemplate
+    public void testSourceParallelism(TestContainer container) throws Exception {
+        LOG.info("=========Multi parallelism testing begins===========");
```
```suggestion
```

Hisoka-X commented on seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java:
```diff
@@ -77,6 +77,8 @@ public void registerReader(int subtaskId) {
         if (assigned < 0) {
             assigned = subtaskId;
             context.assignSplit(subtaskId, new ClickhouseSourceSplit());
+        } else {
+            context.signalNoMoreSplits(subtaskId);
         }
```
```suggestion
        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
            context.signalNoMoreSplits(subtaskId);
        } else {
            context.signalNoMoreSplits(subtaskId);
        }
```

zhilinli123 replied:
>  Write it like this?
 
```

        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
        }
        context.signalNoMoreSplits(subtaskId);

```

Hisoka-X replied:
> > Write it like this?
> 
> ```
> 
>         if (assigned < 0) {
>             assigned = subtaskId;
>             context.assignSplit(subtaskId, new ClickhouseSourceSplit());
>         }
>         context.signalNoMoreSplits(subtaskId);
> ```

+1

YOMO-Lee replied:
Okay, I didn't notice this

Hisoka-X commented on seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java:
```diff
@@ -97,6 +101,12 @@ record -> {
             }
             this.readerContext.signalNoMoreElement();
             this.splits.clear();
+        } else if (noMoreSplit
+                && splits.isEmpty()
+                && Boundedness.BOUNDED.equals(readerContext.getBoundedness())) {
+            log.info("Closed the bounded ClickHouse source");
+            this.readerContext.signalNoMoreElement();
+            this.splits.clear();
         }
```
```suggestion
            this.splits.clear();
        }
        if (noMoreSplit
                && splits.isEmpty()
                && Boundedness.BOUNDED.equals(readerContext.getBoundedness())) {
            log.info("Closed the bounded ClickHouse source");
            this.readerContext.signalNoMoreElement();
            this.splits.clear();
        }
```

zhilinli123 replied:
`Boundedness.BOUNDED.equals(readerContext.getBoundedness())`
This can be removed I think because clickhouse only supports batch mode reading
cc @Hisoka-X

Hisoka-X replied:
Keep it for now. The purpose of keeping it is to unify the processing logic of most connectors, to facilitate reference for other developers and reduce the chance of bugs.

zhilinli123 replied:
> Keep it for now. The purpose of keeping it is to unify the processing logic of most connectors, to facilitate reference for other developers and reduce the chance of bugs.

What if the user uses processing mode to write STREAMING? It's unnecessary code and I think it's better to get rid of it Developers adding features are certainly familiar with their read patterns

Hisoka-X replied:
> What if the user uses processing mode to write STREAMING?

Sorry, I don get it. Could you share more details? You meaning set `job.mode` to `STREAMING`?

zhilinli123 replied:
> > What if the user uses processing mode to write STREAMING?
> 
> Sorry, I don get it. Could you share more details? You meaning set `job.mode` to `STREAMING`?
```

env {
  job.mode = "STREAMING"
  parallelism = 2
}
source {
  # This is a example source plugin **only for test and demonstrate the feature source plugin**
  Clickhouse {
    host = "127.0.0.1:8123"
    database = "default"
    sql = "select * from students"
    username = "default"
    password = ""
  }
  # If you would like to get more information about how to configure seatunnel and see full list of source plugins,
  # please go to https://seatunnel.apache.org/docs/connector-v2/source/ClickhouseSource
}


sink {
  Console {
  }
  # If you would like to get more information about how to configure seatunnel and see full list of sink plugins,
  # please go to https://seatunnel.apache.org/docs/connector-v2/sink
}

```

So for example in this case the user will never exit the program, essentially no matter how you configure the read mode clickhouse only supports batch mode, right

Hisoka-X replied:
If users set to STREAMING, but source only supported BATCH mode. The job will execute as BATCH mode. It depends on https://github.com/apache/seatunnel/blob/a8d0d4ce77a1ac9116f93381dfe296d22d8f0e8b/seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSource.java#L160

zhilinli123 replied:
Understood, but an extra judgment is a redundant operation

Hisoka-X reviewed (commented):
Please add checkpoint lock when read spilt. Please refer https://github.com/apache/seatunnel/blob/50113e77cf6729c833bca6a5e20ef397b3992782/seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/source/reader/DorisSourceReader.java#L76

Hisoka-X commented:
cc @zhilinli123 as well.

zhilinli123 commented:
> cc @zhilinli123 as well.
1.Thanks for ping me, I think pom needs to be added 
```

   <dependency>
            <groupId>org.apache.seatunnel</groupId>
            <artifactId>seatunnel-api</artifactId>
            <version>${project.version}</version>
        </dependency>
```

YOMO-Lee commented:
> > cc @zhilinli123 as well.
> > 1.Thanks for ping me, I think pom needs to be added
> 
> ```
> 
>    <dependency>
>             <groupId>org.apache.seatunnel</groupId>
>             <artifactId>seatunnel-api</artifactId>
>             <version>${project.version}</version>
>         </dependency>
> ```

I don't quite understand where this should be added and where it needs to be used. Currently, I haven't found any places where it needs to be used

zhilinli123 commented:
> > > cc @zhilinli123 as well.
> > > 1.Thanks for ping me, I think pom needs to be added
> > 
> > 
> > ```
> > 
> >    <dependency>
> >             <groupId>org.apache.seatunnel</groupId>
> >             <artifactId>seatunnel-api</artifactId>
> >             <version>${project.version}</version>
> >         </dependency>
> > ```
> 
> I don't quite understand where this should be added and where it needs to be used. Currently, I haven't found any places where it needs to be used

Boundedness.BOUNDED  quoted https://github.com/YOMO-Lee/seatunnel/blob/clickhouse/seatunnel-connectors-v2/connector-clickhouse/pom.xml

zhilinli123 commented:
Waiting for the start of the current pr merger modify this code: https://github.com/apache/seatunnel/pull/7529

YOMO-Lee committed:
[Fix][Connector-V2] Fixed clickhouse connectors cannot stop under multiple parallelism

Adjust the bug regarding the Clickhouse connector not being able to stop properly

zhilinli123 commented on seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java:
```diff
@@ -75,28 +79,35 @@ public void close() throws IOException {
+                }
+                this.readerContext.signalNoMoreElement();
+                this.splits.clear();
+            }
```
Extract clickhouse Close the task method
```
                this.readerContext.signalNoMoreElement();
                this.splits.clear();
```

zhilinli123 commented on seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java:
```diff
@@ -75,28 +79,35 @@ public void close() throws IOException {
+            }
+            if (noMoreSplit && splits.isEmpty()) {
+                log.info("Closed the bounded ClickHouse source");
+                this.readerContext.signalNoMoreElement();
```
ditto

YOMO-Lee replied:
OK

zhilinli123 commented on seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java:
```diff
@@ -77,6 +77,9 @@ public void registerReader(int subtaskId) {
             assigned = subtaskId;
             context.assignSplit(subtaskId, new ClickhouseSourceSplit());
+            context.signalNoMoreSplits(subtaskId);
+        } else {
```
```

        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
        }
        context.signalNoMoreSplits(subtaskId);
```

YOMO-Lee committed:
[Fix][Connector-V2] Fixed clickhouse connectors cannot stop under multiple parallelism

Optimize code structure

YOMO-Lee committed:
[Fix][Connector-V2] Fixed clickhouse connectors cannot stop under multiple parallelism

format code

zhilinli123 commented:
Wait for CI pass

Hisoka-X reviewed (approved):
LGTM if ci passes. Thanks @YOMO-Lee and @zhilinli123 @corgy-w for review!

zhilinli123 commented:
> LGTM if ci passes. Thanks @YOMO-Lee and @zhilinli123 @corgy-w for review!

+1

hailin0 commented:
waiting for ci passed
